xmtp_api_grpc_gateway/src/lib.rs

```diff
@@ -36,67 +47,153 @@ impl XmtpApiClient for XmtpGrpcGatewayClient {
+        Ok(response
+            .json::<QueryResponse>()
+            .await
+            .map_err(|e| Error::new(ErrorKind::QueryError).with(e))?)
     }
 
     async fn batch_query(&self, request: BatchQueryRequest) -> Result<BatchQueryResponse, Error> {
-        let res: BatchQueryResponse = self
+        let response = self
             .http
             .post(&format!("{}/message/v1/batch-query", self.url))
             .json(&request)
             .send()
             .await
-            .map_err(|e| Error::new(ErrorKind::BatchQueryError).with(e))?
-            .json()
-            .await
             .map_err(|e| Error::new(ErrorKind::BatchQueryError).with(e))?;
-        Ok(res)
+        Ok(response
+            .json::<BatchQueryResponse>()
+            .await
+            .map_err(|e| Error::new(ErrorKind::BatchQueryError).with(e))?)
+    }
+}
+
+pub struct XmtpGrpcGatewaySubscription<'a> {
+    // When this is `None`, the stream has been closed.
+    pub stream: Option<LocalBoxStream<'a, Result<Envelope, Error>>>,
+}
+
+// The result of calling .subscribe()
+// The grpc-gateway streams newline-delimited JSON bodies
+// in this shape:
+//  { result: { ... Envelope ... } }\n
+//  { result: { ... Envelope ... } }\n
+//  { result: { ... Envelope ... } }\n
+// So we use this to pluck the `Envelope` as the `result`.
+#[derive(Deserialize, Debug)]
+struct SubscribeResult {
+    result: Envelope,
+}
+
+impl<'a> XmtpGrpcGatewaySubscription<'a> {
+    pub fn start(req: LocalBoxStream<'a, Result<Bytes, reqwest::Error>>) -> Self {
+        let bytes_reader = StreamReader::new(
+            req.map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err)),
+        );
+        let codec = LinesCodec::new_with_max_length(1024 * 1024);
+        let frames_reader = FramedRead::with_capacity(bytes_reader, codec, 8 * 1024 * 1024);
+        let stream = frames_reader
+            .map(|frame_res| match frame_res {
+                Ok(frame_str) => serde_json::from_str::<SubscribeResult>(frame_str.as_str())
+                    .map(|v| v.result)
+                    .map_err(|e| Error::new(ErrorKind::SubscribeError).with(e)),
+                Err(err) => Err(Error::new(ErrorKind::SubscribeError).with(err)),
+            })
+            .boxed_local();
+
+        XmtpGrpcGatewaySubscription {
+            stream: Some(stream),
+        }
     }
 }
 
-// TODO: implement JSON segmented streaming
-pub struct XmtpGrpcGatewaySubscription {}
-impl XmtpApiSubscription for XmtpGrpcGatewaySubscription {
+#[async_trait(?Send)]
+impl<'a> XmtpApiSubscription for XmtpGrpcGatewaySubscription<'a> {
     fn is_closed(&self) -> bool {
-        true
+        return self.stream.is_none();
     }
 
-    fn get_messages(&self) -> Vec<Envelope> {
-        vec![]
+    // HACK: this consumes from the stream whatever is already ready.
+    // TODO: implement a JS-friendly promise/future interface instead
+    async fn get_messages(&mut self) -> Vec<Envelope> {
```

Comment:
After much-mashing to get this working, I think we should scrap this and change the abstraction for `XmtpApiSubscription` across the library.  I suspect it will be cleaner either as a proper `Stream` (i.e. ✂️  the `get_messages()` stuff and use `.collect()` and other stream tools) and/or by designing outside-in to use JS promises (qua rust `Futures`) and callbacks.

Reply:
I have no objections to this